### PR TITLE
Implement remote file search

### PR DESCRIPTION
**Pull request by gerritholl**
Implement searching for files remotely by accepting any object implementing a ``.glob(...)`` method, while defaulting to the `LocalFileSystem` class from the fsspec module, which is now a new dependency.  This enables searching for files from S3 and many other remote file systems.

 - [ ] Closes #xxxx 
 - [x] Tests added 
 - [x] Tests passed 
 - [x] Passes ``flake8 satpy`` 
 - [x] Fully documented 
 - [x] Add your name to `AUTHORS.md` if not there already

### CODE REVIEW

**Comment by gerritholl** [comment]
I've currently settled to fall back on `glob` if `fs` is `None`, regardless of whether `fsspec` is installed or not, because I found there are behaviour differences between `glob.iglob` (or `glob.glob`) and `LocalFileSystem.glob` (see earlier test failured for details). It would be a bit tricky to test for those differences in a unit test, as I don't know how to set up a test environment to pretend a module is _not_ installed (I suppose it can be done, I just don't know how).  With the latest approach (38f03de), there should be full backward compatibility.

**Comment by gerritholl** [comment]
I believe the OSX test failures are unrelated to this PR.

**Comment by mraspaud** [comment]
@gerritholl looks like the rasterio problem @djhoese mentionned in another PR

**Comment by djhoese** [comment]
@gerritholl One way to do it would be something like:

```
try:
    import fsspec
except ImportError:
    fsspec = None
```

In the module that is being tested. Then in the function do `if fsspec is None`. Then in the tests you can force the module to not be installed by doing:

```
import satpy.readers
with mock.patch.object(satpy.readers, 'fsspec', None):
    # your test code
```

At least I think that should work.

FYI for the failing rasterio stuff: https://github.com/conda-forge/rasterio-feedstock/issues/161

**Comment by gerritholl** [comment]
It doesn't require any extras.  The fsspec is needed for the tests, but the code itself can use any class that has a `glob` method.

**Comment by gerritholl** [comment]
I don't understand why coveralls claims this PR reduces coverage.

**Comment by gerritholl** [comment]
Despite the codecov comment being updated an hour ago, it appears to ignore the three most recent commits.  Its claim of 18 coverage changes is clearly wrong.  [recent commits are missing](https://codecov.io/gh/pytroll/satpy/pull/1169/commits) and sure, comparing a 19 day old branch commit against current master will show a drop in coverage, but it's not comparing the commits I pushed today (which include merging the current master branch).

File: satpy/readers/__init__.py [hunk 1]
```diff
@@ -582,17 +582,26 @@ def available_readers(as_dict=False):
 def find_files_and_readers(start_time=None, end_time=None, base_dir=None,
                            reader=None, sensor=None, ppp_config_dir=None,
                            filter_parameters=None, reader_kwargs=None,
-                           missing_ok=False):
-    """Find on-disk files matching the provided parameters.
-
+                           missing_ok=False, fs=None):
+    """Find files matching the provided parameters.
+
```
Comment (mraspaud) [inline satpy/readers/__init__.py:587]:
```suggestion

```

**Pull request merged**